larpandoracontent/LArControlFlow/BdtBeamParticleIdTool.cc

```diff
@@ -0,0 +1,759 @@
+
+            const double separationOneNu((interceptOneNu - m_pSliceFeatureParameters->GetBeamLArTPCIntersection()).GetMagnitude());
+            const double separationTwoNu((interceptTwoNu - m_pSliceFeatureParameters->GetBeamLArTPCIntersection()).GetMagnitude());
+
+            separationNu = std::min(separationOneNu, separationTwoNu);
+
+            // Cosmic
+            CartesianVector centroidCr(0.f, 0.f, 0.f);
+            LArPcaHelper::EigenVectors eigenVecsCr;
+            LArPcaHelper::RunPca(selectedCaloHitListCr, centroidCr, eigenValuesCr, eigenVecsCr);
+
+            const CartesianVector &majorAxisCr(eigenVecsCr.front());
+            supplementaryAngleToBeamCr = majorAxisCr.GetOpeningAngle(m_pSliceFeatureParameters->GetBeamDirection());
+
+            CartesianVector interceptOneCr(0.f, 0.f, 0.f), interceptTwoCr(0.f, 0.f, 0.f);
+            this->GetLArTPCIntercepts(centroidCr, majorAxisCr, interceptOneCr, interceptTwoCr);
+
+            const double separationOneCr((interceptOneCr - m_pSliceFeatureParameters->GetBeamLArTPCIntersection()).GetMagnitude());
+            const double separationTwoCr((interceptTwoCr - m_pSliceFeatureParameters->GetBeamLArTPCIntersection()).GetMagnitude());
+
+            separationCr = std::min(separationOneCr, separationTwoCr);
+
+            float maxYNu(-std::numeric_limits<float>::max()), maxYCr(-std::numeric_limits<float>::max());
+
+            for (const CaloHit *pCaloHit : caloHitList3DNu)
+            {
+                if (maxYNu < pCaloHit->GetPositionVector().GetY())
+                    maxYNu = pCaloHit->GetPositionVector().GetY();
+            }
+
+            for (const CaloHit *pCaloHit : caloHitList3DCr)
+            {
+                if (maxYCr < pCaloHit->GetPositionVector().GetY())
+                    maxYCr = pCaloHit->GetPositionVector().GetY();
+            }
+
+            m_featureVector.push_back(closestDistanceNu);
+            m_featureVector.push_back(supplementaryAngleToBeamNu);
+            m_featureVector.push_back(separationNu);
+            m_featureVector.push_back(eigenValuesNu.GetX());
+            m_featureVector.push_back(eigenValuesNu.GetY());
+            m_featureVector.push_back(eigenValuesNu.GetZ());
+            m_featureVector.push_back(maxYNu);
+            m_featureVector.push_back(allConnectedPfoListNu.size());
+
+            m_featureVector.push_back(closestDistanceCr);
+            m_featureVector.push_back(supplementaryAngleToBeamCr);
+            m_featureVector.push_back(separationCr);
+            m_featureVector.push_back(eigenValuesCr.GetX());
+            m_featureVector.push_back(eigenValuesCr.GetY());
+            m_featureVector.push_back(eigenValuesCr.GetZ());
+            m_featureVector.push_back(maxYCr);
+            m_featureVector.push_back(allConnectedPfoListCr.size());
+
+            m_isAvailable = true;
+        }
+    }
+    catch (StatusCodeException &)
```

Comment:
I believe this is well understood.  The general idea being if a variable can't be set for the given inputs then m_isAvailable is set to false and the slice is classed as a cosmic ray.  There's no point doing mini try-catch blocks for each variable because all variables are needed for the bdt to work.  Therefore, as you can't proceed if any of the calculations fail a large try catch block seems appropriate.